File: src/main/java/io/pageobject/generator/element/Elements.java
package io.pageobject.generator.element;

import com.google.common.base.Strings;
import io.pageobject.generator.GeneratorContext;
import org.jsoup.helper.StringUtil;
import org.jsoup.nodes.Element;

import static io.pageobject.generator.attribute.Attributes.hasNormalizedAttribute;

public class Elements {

    public static boolean isButton(Element element) {
        return element.tagName().equals("button") || (element.tagName().equals("input") &&
                                                      (element.attr("type").equals("button") ||
                                                       element.attr("type").equals("submit")));
    }

    public static boolean isLink(Element element) {
        return element.tagName().equals("a");
    }

    public static String generateCssSelector(Element element, GeneratorContext context) {
        if (element.id().length() > 0) {
            return "#" + element.id();
        }

        String tagName = element.tagName().replace(':', '|');
        StringBuilder selector = new StringBuilder(tagName);
        addClasses(element, selector);

        Element parent = element.parent();
        if (parent == null) {
            return selector.toString();
        }

        if (parent.tagName().equals("body")) {
            if (context.getSource().contains("<body")) {
                return "body > " + selector.toString();
            } else {
                return selector.toString();
            }
        }

        String directChildrenSelector = ">" + selector.toString();
        int numberOfDirectChildren = getNumberOfChildren(directChildrenSelector, parent);

        if (context.isNestedElement()) {
            int numberOfAllChildren = getNumberOfChildren(selector.toString(), parent);
            if (numberOfAllChildren != numberOfDirectChildren) {
                return null;
            }
        }

        if (numberOfDirectChildren > 1) {
            org.jsoup.select.Elements childrenWithSameSelector = parent.select(directChildrenSelector);
            String type = childrenWithSameSelector.get(0).tagName();
            org.jsoup.select.Elements childrenWithSameType = parent.select(">" + type);
            int siblingIndex = childrenWithSameType.indexOf(element);

            if (anyPrecedingSiblingIsAStopElement(childrenWithSameType, siblingIndex)) {
                return null;
            }

            selector.append(String.format(":nth-of-type(%d)", siblingIndex + 1));
        }

        if (hasReachedLastRepeaterElement(context, parent)) {
            return selector.toString();
        }

        selector.insert(0, " > ");

        String parentSelector = generateCssSelector(parent, context);
        return parentSelector == null ? null : parentSelector + selector.toString();
    }

    private static void addClasses(Element element, StringBuilder selector) {
        String classes = StringUtil.join(element.classNames(), ".");
        if (classes.length() > 0) {
            selector.append('.').append(classes);
        }
    }

    private static int getNumberOfChildren(String selector, Element parent) {
        org.jsoup.select.Elements elements = parent.select(selector);
        elements.remove(parent);
        return elements.size();
    }

    private static boolean anyPrecedingSiblingIsAStopElement(org.jsoup.select.Elements childrenWithSameType,
                                                             int siblingIndex) {
        for (int i = 0; i <= siblingIndex; i++) {
            Element sibling = childrenWithSameType.get(i);
            if (isStopElement(sibling)) {
                return true;
            }
        }
        return false;
    }

    private static boolean isStopElement(Element element) {
        String[] normalizedAttributes = new String[] {"ng-if", "ng-repeat"};

        for (String normalizedAttribute : normalizedAttributes) {
            if (hasNormalizedAttribute(element, normalizedAttribute)) {
                return true;
            }
        }

        String[] regularAttributes = new String[] {"*ngIf", "*ngFor"};

        for (String regularAttribute : regularAttributes) {
            if (!Strings.isNullOrEmpty(element.attr(regularAttribute))) {
                return true;
            }
        }

        return false;
    }

    private static boolean hasReachedLastRepeaterElement(GeneratorContext context, Element parent) {
        return context.isNestedElement() && context.getLastRepeaterElement() == parent;
    }

}

File: lib/element/elements.js
'use strict';

const getAttribute = require('../dom-helpers').getAttribute;
const hasNormalizedAttribute = require('../attribute/attributes').hasNormalizedAttribute;
const _ = require('lodash');
const select = require('css-select');

class Elements {

  static isButton(element) {
    return element.tagName === 'button' || (element.tagName === 'input' &&
                                            (getAttribute(element, 'type') === 'button' ||
                                             getAttribute(element, 'type') === 'submit'));
  }

  static isLink(element) {
    return element.tagName === 'a';
  }

  static generateCssSelector(element, context) {
    let id = getAttribute(element, 'id');
    if (!_.isEmpty(id)) {
      return '#' + id;
    }

    let tagName = element.tagName.replace(/:/g, '|');
    let selector = tagName;
    selector = addClasses(element, selector);

    let parent = element.parent;
    if (parent.tagName === 'root') {
      return selector;
    }

    if (parent.tagName === 'body') {
      if (context.source.includes('<body')) {
        return 'body > ' + selector;
      } else {
        return selector;
      }
    }

    let directChildrenSelector = '>' + selector;
    let numberOfDirectChildren = getNumberOfChildren(directChildrenSelector, parent);

    if (context.isNestedElement()) {
      let numberOfAllChildren = getNumberOfChildren(selector, parent);
      if (numberOfAllChildren !== numberOfDirectChildren) {
        return null;
      }
    }

    if (numberOfDirectChildren > 1) {
      let childrenWithSameSelector = select(directChildrenSelector, parent);
      let type = childrenWithSameSelector[0].tagName;
      let childrenWithSameType = select('>' + type, parent);
      let siblingIndex = childrenWithSameType.indexOf(element);

      if (anyPrecedingSiblingIsAStopElement(childrenWithSameType, siblingIndex)) {
        return null;
      }

      selector += `:nth-of-type(${siblingIndex + 1})`;
    }

    if (hasReachedLastRepeaterElement(context, parent)) {
      return selector;
    }

    selector = ' > ' + selector;

    let parentSelector = Elements.generateCssSelector(parent, context);
    return parentSelector == null ? null : parentSelector + selector;
  }

  static classList(element) {
    let classAttrValue = getAttribute(element, 'class');
    return !_.isEmpty(classAttrValue) ? classAttrValue.trim().split(/\s+/g) : [];
  }

}

function addClasses(element, selector) {
  let classes = Elements.classList(element).join('.');
  return classes.length > 0 ? selector + '.' + classes : selector;
}

function getNumberOfChildren(selector, parent) {
  let matches = select(selector, parent);
  return matches.length;
}

function anyPrecedingSiblingIsAStopElement(childrenWithSameType, siblingIndex) {
  for (let i = 0; i <= siblingIndex; i++) {
    let sibling = childrenWithSameType[i];
    if (isStopElement(sibling)) {
      return true;
    }
  }
  return false;
}

function isStopElement(element) {
  let normalizedAttributes = ['ng-if', 'ng-repeat'];

  for (let normalizedAttribute of normalizedAttributes) {
    if (hasNormalizedAttribute(element, normalizedAttribute)) {
      return true;
    }
  }

  let regularAttributes = ['*ngIf', '*ngFor'];

  for (let regularAttribute of regularAttributes) {
    if (!_.isEmpty(getAttribute(element, regularAttribute))) {
      return true;
    }
  }

  return false;
}

function hasReachedLastRepeaterElement(context, parent) {
  return context.isNestedElement() && context.getLastRepeaterElement() === parent;
}

module.exports = Elements;